{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Ошибки, дата_время.ipynb","provenance":[],"collapsed_sections":[],"toc_visible":true,"authorship_tag":"ABX9TyNENZEOhPgSA2KfAR7Zrn1e"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"C1mhrX78Trm8"},"source":["# Задание 1\n","Печатные газеты использовали свой формат дат для каждого выпуска. Для каждой газеты из списка напишите формат указанной даты для перевода в объект datetime:\n","The Moscow Times - Wednesday, October 2, 2002\n","The Guardian - Friday, 11.10.13\n","Daily News - Thursday, 18 August 1977"]},{"cell_type":"code","metadata":{"id":"TcDEeGPvTuuI"},"source":["The Moscow Times - '%A, %B %d, %Y'\n","The Guardian - '%A, %d.%m.%y'\n","Daily News - '%A, %d %B %Y'"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"wGPfG9j_gTzA"},"source":["# Задание 2\n","Дан поток дат в формате YYYY-MM-DD, в которых встречаются некорректные значения:\n","stream = [‘2018-04-02’, ‘2018-02-29’, ‘2018-19-02’]\n","\n","Напишите функцию, которая проверяет эти даты на корректность. Т. е. для каждой даты возвращает True (дата корректна) или False (некорректная дата)."]},{"cell_type":"code","metadata":{"id":"2SwPCnHjgW9q"},"source":["from datetime import datetime\n","\n","def is_date_correct(dates):\n","  for date in dates:\n","    try:\n","      datetime.strptime(date, '%Y-%m-%d')\n","      print(True) # использовал print, т.к. return прекратил бы выполнение и до проверки других дат не дошло бы\n","    except:\n","      print(False)\n","\n","stream = ['2018-02-29', '2018-04-02', '2018-19-02']\n","is_date_correct(stream)\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"bji8_xosggtU"},"source":["# Задание 3\n","Напишите функцию date_range, которая возвращает список дат за период от start_date до end_date. Даты должны вводиться в формате YYYY-MM-DD. В случае неверного формата или при start_date > end_date должен возвращаться пустой список."]},{"cell_type":"code","metadata":{"id":"H11PBKLwgkEf","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1629481878567,"user_tz":-300,"elapsed":250,"user":{"displayName":"Pavel Novoselov","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gi_XD9mukCXRDvWdlNwxFfx68YDgqYvZR-_BTTGPw=s64","userId":"00991992689762570163"}},"outputId":"2585cabf-e825-4e46-b8a1-39040494bdf7"},"source":["from datetime import datetime\n","from datetime import timedelta\n","\n","def date_range(start_date, end_date):\n","  date_list = []\n","\n","  try:\n","      start_dt = datetime.strptime(start_date, '%Y-%m-%d')\n","      end_dt = datetime.strptime(end_date, '%Y-%m-%d')\n","\n","  except:\n","      print('Что-то пошло не так!')\n","      return []\n","\n","  if start_dt > end_dt:\n","        return []\n","\n","  current_dt = start_dt\n","  while current_dt <= end_dt:\n","     date_list.append(current_dt.strftime('%Y-%m-%d'))\n","     current_dt += timedelta(days=1)\n","\n","  print(date_list)\n","\n","date_range('2020-03-01', '2020-03-08')\n"],"execution_count":3,"outputs":[{"output_type":"stream","text":["['2020-03-01', '2020-03-02', '2020-03-03', '2020-03-04', '2020-03-05', '2020-03-06', '2020-03-07', '2020-03-08']\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"epXHrRRdg4PV"},"source":["# Задание 4\n","Ваш коллега прислал код функции:\n","\n","DEFAULT_USER_COUNT = 3\n","\n","def delete_and_return_last_user(region, default_list=[‘A100’, ‘A101’, ‘A102’]):\n","\"\"“\n","Удаляет из списка default_list последнего пользователя\n","и возвращает ID нового последнего пользователя.\n","”\"\"\n","element_to_delete = default_list[-1]\n","default_list.remove(element_to_delete)\n","\n","1\n","return default_list[DEFAULT_USER_COUNT-2]\n","При однократном вызове этой функции все работает корректно:\n","delete_and_return_last_user(1)\n","‘A101’\n","\n","Однако, при повторном вызове получается ошибка IndexError: list index out of range.\n","\n","Задание:\n","\n","Что значит ошибка list index out of range?\n","Почему при первом запуске функция работает корректно, а при втором - нет?"]},{"cell_type":"code","metadata":{"id":"5EhUBXNJhASk"},"source":["DEFAULT_USER_COUNT = 3\n","\n","def delete_and_return_last_user(region, default_list=['A100', 'A101', 'A102']):\n","  \n","   element_to_delete = default_list[-1] \n","   default_list.remove(element_to_delete)\n","   return default_list[DEFAULT_USER_COUNT-2]\n","\n","delete_and_return_last_user(1)\n","delete_and_return_last_user(1)\n","\n","# Ошибка \"IndexError: list index out of range\" означает, что код обращается к индексу элемента, которого нет в списке\n","# Ошибка возникает из-за того, что значения по умолчанию вычисляются один раз при определении функции, а не при каждом вызове\n","# Поэтому при первом запуске из списка удаляется один элемент и доступные индексы остаются 0 и 1,\n","# при повторном запуске удаляется ещё один элемент и индекс остаётся только 0,\n","# и когда код пытается вернуть default_list[DEFAULT_USER_COUNT-2], то он обращается к индексу 1, которого нет.\n","# Способ решения - использовать по умолчанию None и проверять в теле функции, если параметр не передан, т.е. None, то объявлять список"],"execution_count":null,"outputs":[]}]}